### ISSUE TITLE
Safari - iOS cookie handling doesn't find any secure cookies

### ISSUE THREAD

**Issue by hazmeister** [appium/appium#6797]
## The problem

To reset the browser to a clean state in my tests, I often use the `deleteAllCookies()` method to clear the test's state.  With Appium and iOS, cookie handling appears to be limited to non secure cookies for the same domain.  Even checking the number of cookies after calling `deleteAllCookies()` seems to leave several cookies on sites secured with HTTPS.

Running the same test with Appium and Android deletes all the cookies for the domain as expected.
## Environment
- Appium version (or git revision) that exhibits the issue: 1.53
- Desktop OS/version used to run Appium: Mac OS X 10.11.6
- Mobile platform/version under test: iOS 9.3
- Real device or emulator/simulator: Both
- Appium CLI or Appium.app|exe: Appium.app
## Code To Reproduce Issue

Step through in debug mode - subsititue your own account details!

```
        driver.get("https://accounts.google.com/AccountChooser");
        driver.findElement(By.cssSelector("#Email")).sendKeys("youraccount@gmail.com");
        driver.findElement(By.cssSelector("input#next")).click();
        driver.findElement(By.cssSelector("#Passwd")).sendKeys("password");
        driver.findElement(By.cssSelector("input#signIn")).click();

        int clear = 0;
        int secure = 0;
        Set<Cookie> cookies = driver.manage().getCookies();
        for (Cookie cookie : cookies) {
            if(cookie.isSecure()) {
                secure++;
            } else {
                clear++;
            }
        }
        System.out.println("Found " + secure + " secure cookies");
        System.out.println("Found " + clear + " clear cookies");
        driver.manage().deleteAllCookies();
        driver.navigate().refresh();
        Set<Cookie> cookies1 = driver.manage().getCookies();
        System.out.println(cookies1.size() + " still found.";
```

[Clone here](https://github.com/hazmeister/appium-bugs/blob/master/src/test/java/Cookies.java)
## Appium Logs

[Gist](https://gist.github.com/hazmeister/33bfcec6ecd9f79f69f6c24b4b1d166a)

**Comment by jlipps** [comment]
Thanks for the repro @hazmeister we'll look into it.

**Comment by hazmeister** [comment]
@jlipps had another look at this with Appium 1.6.3 and the same behaviour persists.  I poked around the appium code a tad - would I be right in thinking that the current implementation uses javascript  to get cookie info from the page?  The secure cookie info appears to be supported by the iOS-webkit-debug-proxy from both the simulator and real devices, so I assume fetching secure cookies should be technically possible. 

Would you be able to point me in the direction of any existing features in Appium which make use of iOS-webkit-debug-proxy?  I might be able to put 2 and 2 together to make nearly 4.

**Comment by jlipps** [comment]
@hazmeister it'd be awesome if you wanted to look into it! I believe you're right that we currently use javascript to retrieve cookies. All of our logic for connecting to the remote debugger is in https://github.com/appium/appium-remote-debugger. Note that we do not use ios-webkit-debug-proxy for sims, we talk to the Safari remote debugger directly. But if it is a part of the remote debug API in general, perhaps a general solution is possible here.

**Comment by eelessam** [comment]
Any progress on this guys? Or anything I can try to help with?

**Comment by imurchie** [comment]
The core team has not had a chance to work on this yet, though it is in our backlog. I haven't seen any action from anyone else, either. :(

**Comment by hazmeister** [comment]
I did dig into this for a while, and found myself on a bit goose chase as the WebKit remote debug documentation seems to have moved home and one more of the resource files from the documentation has since been refactored and moved elsewhere. I'll dig out what I've found and start a thread on appium discuss and link it here.

**Comment by hazmeister** [comment]
Appium's cookie handling for iOS on real devices and simulators is implemented using Javascript.  At first, this appears to work well, but JavaScript is limited to retrieving and editing cookies without the Secure or HTTPOnly flags set.  It is, however, possible to view these cookies when using the Safari Developer Tools, so I'm assuming will be available over the WebKit Debug Protocol.

A quick google lead to [here](https://webkit.org/blog/1875/announcing-remote-debugging-protocol-v1-0/), which although looked promising, but is no longer live on the webkit.org.  Thankfully the wayback machine [has us covered](http://web.archive.org/web/20120411220639/http://www.webkit.org/blog/1875/announcing-remote-debugging-protocol-v1-0).  One good looking link references the JSON schema for the remote debug protocol, which again, is no longer accessible via the original link.  However, the link was to a file in WebKit's source control repo, so I was able to dig back to earlier release to find where the file was still in its [original location](https://trac.webkit.org/browser/releases/WebKitGTK/webkit-2.0/Source/WebCore/inspector/Inspector.json).  I assume the file probably still exists in the later webkit releases, and may have been moved somewhere.

So the next piece of the puzzle is to substitute the existing behaviour for one using the remote debug protocol.  In theory this should work for both sim and real device users.  My latest challenge was trying to find a call to the remote debugger to copy as a base for the new cookie behaviour.  This is where I got stuck.
